### DISCUSSION TITLE
Script to Import PotPlayer Bookmarks into Freeplane

### DISCUSSION THREAD

**Discussion by aaa1386** [freeplane/freeplane#2164] [Help]
I’m looking for a way to import bookmarks created in PotPlayer into Freeplane and organize them as nodes in a mind map. My goal is to categorize and manage these bookmarks more effectively.

My Scenario:
PotPlayer: I use this media player and add bookmarks for video and audio files.
I can create a list of bookmarks and the start time of those bookmarks as a text file, for example:
Intro - 00:00:30
Chapter 1 - 00:05:10
Chapter 2 - 00:12:45
Freeplane: I want a script that reads this text file and creates a new node in  map for each bookmark. For example:
Root Node: Video Title
Child Node: Intro (00:00:30)
Child Node: Chapter 1 (00:05:10)
...

Play a desired bookmark within the map. I mean play it exactly from the point where the bookmark is located in the audio file.
There should be the ability to execute any bookmark within the map. I mean, it should be executed exactly from the point where the bookmark is located in the audio file.
If we put the bookmark nodes in a new category, it should still be possible to execute them.

This was my scenario. You may have a different scenario for managing bookmarks. I would appreciate any guidance.

If anyone has experience with something similar or has written a similar script, I’d greatly appreciate your guidance or an example script. Any help, even pointing me to relevant resources or documentation, would be incredibly valuable. 🙏

Thank you so much for your time and for sharing your knowledge! 🌟

**Comment by macmarrum** [comment]
I think that you can simply copy the text and paste it in Freeplane. Each line will create a new child node.

```text
Intro - 00:00:30
Chapter 1 - 00:05:10
Chapter 2 - 00:12:45
```

If you need to have the time in round brackets, you can use `Edit->Find->Find and replace…`
* Find `- ([^()]+)`  [x] Regular expression
* Replace `($1)`  [x] Regular expression

**Reply by aaa1386** [reply]
> Intro - 00:00:30
> Chapter 1 - 00:05:10
> Chapter 2 - 00:12:45

Thanks for your response. I want the audio to start playing from the time specified for it when I click on the node text. It is not enough to just have a bookmark category of audio files but not be able to run it from within Freeplane.

**Reply by macmarrum** [reply]
I think a simple script to launch PotPlayer on a selected node should do the trick. You could assign it a hot key, e.g. `F5`. So you would select your bookmark node, press `F5` and the file would play in PotPlayer starting from the bookmarked position.

If you upload a sample mind map with your bookmarks and links to your audio/video file(s), I can create such a script for you.

**Reply by aaa1386** [reply]
<p><strong>In discussion #2166</strong>, several solutions were proposed. My goal is to organize and manage audio bookmarks within Freeplane in a way that allows seamless categorization and movement of bookmarks across different categories or nodes.</p>

<p><strong>Look at the picture below:</strong></p>
<p><img src="https://github.com/user-attachments/assets/65824546-7c2c-4d8d-8a5b-0517d971f2fd" alt="image"></p>

<p><strong>Bookmarks (start points) within child nodes</strong> are not independent of their parent node. As a result, if a child node is moved to a different category, the functionality to play the linked audio from its designated start point breaks. <span style="color:gray;">I do not want this scenario.</span></p>

<p><strong>I am looking for a solution similar to the design below:</strong></p>
<p>Each audio start point is fully independent and stored in its own node. This ensures that the node can be moved anywhere in Freeplane without losing its functionality to play the linked audio from the specified start point.</p>
<p>The best case would be for the FP developer to create the ability to play audio from a specific point, just as they have created the ability to play audio and video files from the beginning within FP. <span style="color:gray;">I have created page #2166 for this request.</span></p>

<p><strong>Look at the picture below:</strong></p>
<p><img src="https://github.com/user-attachments/assets/e69364cd-8172-4aec-bbfc-7051d0e3e470" alt="image"></p>

<p>In the diagram (image), the "AAA" start point is completely independent, making it relocatable across the Freeplane map without issues.</p>

<p><strong>The goal is to register a node with audio specifications</strong>, such as the path of that audio file and the start time, so the node has the ability to play that audio file from the specific time.</p>

<p>Thank you for wanting to write a script:</p>
<p><strong>Naturally, I need to link an audio or video file to a node.</strong> By linking to the audio file, I introduce the file path to the node. Inside the node, I would like to edit the starting point and specify a name for the audio playback starting point.</p>
<p>For example: <code>aaa : 02:35</code>.</p>
<p>Write a script such that:
<ul>
  <li>If a link is given in a node without specifying a starting point, clicking on the link will act like a normal link and play the file from the beginning.</li>
  <li>If a time is recorded in that node, such as <code>aaa : 02:35</code>, clicking on the link of that node will start playback from that specific time.</li>
  <li>Once the script is running, it should remain active without needing to be re-executed for each audio file, until the script is turned off manually.</li>
</ul>
</p>

The script should be written in such a way that it is easy to edit and provide data to the nodes. For example, the simplest way is to have it look like this:      aaa, 02:35
I work with PotPlayer,  whenever I open a link from within Node, it runs with PotPlayer. Of course, it would be nice if the script could run with other player, but as long as it can be compatible with PotPlayer, that's enough for me.
I think the explanation is sufficient. If you have any questions, please ask.

**Reply by macmarrum** [reply]
You can try this script: https://github.com/macmarrum/freeplane-scripting/raw/refs/heads/main/src/scripts/openWithPotplayer.groovy

You'll need to select a "bookmark" node and run the script on it.

**Reply by aaa1386** [reply]
> You can try this script: https://github.com/macmarrum/freeplane-scripting/raw/refs/heads/main/src/scripts/openWithPotplayer.groovy

It doesn't work.
https://github.com/user-attachments/assets/f458e2e0-8fa4-44b1-8a70-39bd17be96da

**Reply by macmarrum** [reply]
I can see that PotPlayer is launched and tries to play the file. I don't know why it doesn't. You might need to ask in the PotPlayer forums.

You can try a different player. On my Linux laptop I use [mpv](http://mpv.io). I know there is an "enhanced" [mpv.net](https://github.com/mpvnet-player/mpv.net) for Windows.

The following script works for me:
https://github.com/macmarrum/freeplane-scripting/raw/refs/heads/main/src/scripts/openWithMpv.groovy

You might need to adjust `mpvPath` in the script to match your system.

**Comment by aaa1386** [comment]
Thanks everyone
I was able to achieve what I wanted with this script

<img width="778" height="273" alt="image" src="https://github.com/user-attachments/assets/16144c6a-f55e-4451-a2e6-e201a2d351ab" />

**Reply by aaa1386** [reply]
```
import java.net.URLDecoder

def timeNode = node
def parentNode = timeNode.parent

def link = parentNode.link?.text ?: ""
if (!link) {
    ui.errorMessage("لینک فایل در گره پدر وجود ندارد.")
    return
}

def mediaPath = link.replaceFirst(/^file:\/+/, "")
mediaPath = URLDecoder.decode(mediaPath, "UTF-8")

def file = new File(mediaPath)
if (!file.exists()) {
    ui.errorMessage("فایل یافت نشد:\n" + mediaPath)
    return
}

def timeText = timeNode.text.trim()
def totalSeconds = 0

if (timeText =~ /^\d{1,2}m\d{1,2}s$/) {
    def matcher = (timeText =~ /(\d+)m(\d+)s/)
    if (matcher.find()) {
        totalSeconds = matcher.group(1).toInteger() * 60 + matcher.group(2).toInteger()
    }
} else if (timeText =~ /^\d{1,2}:\d{2}$/) {
    def parts = timeText.split(":")
    totalSeconds = parts[0].toInteger() * 60 + parts[1].toInteger()
} else {
    ui.errorMessage("زمان باید به شکل 2m30s یا 02:30 وارد شود.")
    return
}

def potplayerPath = "C:\\Program Files\\DAUM\\PotPlayer\\PotPlayerMini64.exe"

def command = [
    "cmd", "/c",
    "start", "",
    "\"${potplayerPath}\"",
    "\"${mediaPath}\"",
    "/seek=${totalSeconds}"
]

def process = command.execute()
process.waitFor()

```
